File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline

k, n = map(int, input().split())
line = [int(input()) for _ in range(k)]
line.sort()

def bs(array, start, end, target):
    while start <= end:
        mid = (start+end)//2
        cnt = 0
        for i in array:
            cnt += i//mid

        # 만들어진 랜선개수가 필요한 개수보다 많으면 랜선길이를 늘려서 만들어질 개수를 줄여야 함
        if cnt >= target: 
            start = mid + 1
        else:
            end = mid - 1
    return end

print(bs(line, 1, line[k-1], n))

File: 백준/Silver II/1654. 랜선 자르기/랜선 자르기.java
import java.util.*;
import java.io.*;

/*
 * 최대 랜선 길이 구하기
 * N개의 같은 길이의 랜선으로 만들자!
 */
public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int k = Integer.parseInt(st.nextToken());
		int n = Integer.parseInt(st.nextToken());
		
		int[] arr = new int[k];
		for (int i=0; i<k; i++) {
			arr[i] = Integer.parseInt(br.readLine());
		}
		
		Arrays.sort(arr);
		
		sb.append(binarySearch(arr, 1, arr[k-1], n));
		System.out.println(sb);
	}
	
	public static long binarySearch(int[] arr, long start, long end, int target) {
		long result = 0;
		while (start <= end) {
			long mid = (start+end)/2;
			long cnt = 0;
			for (int i: arr) {
				cnt += i/mid;
			}
			
			if (cnt >= target) {
				start = mid+1;
				result = mid;
			} else {
				end = mid-1;
			}
		}
		return result;
	}
}
